{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Determine burst sizes from single-cell fluorescent data\n",
    "\n",
    "Data from the seqFISH (transcriptomic fluorescence assays) paper \"High-resolution spatial multi-omics reveals cell-type specific nuclear compartments\" by Takei *et al.* (2023).\n",
    "\n",
    "Preprint: https://www.biorxiv.org/content/10.1101/2023.05.07.539762v1\n",
    "\n",
    "Raw Data: https://outlook.office.com/mail/safelink.html?url=https://zenodo.org/record/7693825&corid=cddd4553-1e93-1bbb-63f9-c163d915e8dc "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions \n",
    "\n",
    "def get_burst_size_per_cell(count_df):\n",
    "    ''' Return cell by gene count matrix with burst sizes, which is the average of intensity of intronic spots\n",
    "        observed per gene per cell. \n",
    "    '''\n",
    "    \n",
    "    num_genes = len(count_df['geneID'].unique())\n",
    "    num_cells = len(count_df['unique_id'].unique())\n",
    "    df_ = count_df.groupby(['unique_id','geneID'])['intensity'].mean().reset_index(name='mean_burst_size')\n",
    "    \n",
    "    genes = count_df['geneID'].unique()\n",
    "    cells = count_df['unique_id'].unique()\n",
    "    genes_long = list(genes)*len(cells)\n",
    "\n",
    "    cells_long = []\n",
    "\n",
    "    for c in cells:\n",
    "        cells_long+=[c]*len(genes)\n",
    "    \n",
    "    all_genes_cells = pd.DataFrame({'geneID': genes_long,'unique_id' : cells_long})\n",
    "    df_merge = all_genes_cells.merge(df_,left_on=['geneID','unique_id'],right_on=['geneID','unique_id'],\n",
    "                                how='left').fillna(0)\n",
    "    \n",
    "    count_array = np.zeros((num_cells,num_genes))\n",
    "    \n",
    "    for i,ID in enumerate(count_df['unique_id'].unique()):\n",
    "        counts_ = df_merge[df_merge['unique_id']==ID]\n",
    "        count_array[i,:] = counts_['mean_burst_size']\n",
    "    \n",
    "        \n",
    "    return(count_array)\n",
    "\n",
    "\n",
    "def get_count_matrix(count_df):\n",
    "    ''' Given a count data frame with each count, return cell by gene matrix \n",
    "    with number of times each gene was observed in each cell (not weighted by intensity).\n",
    "    '''\n",
    "    \n",
    "    num_genes = len(count_df['geneID'].unique())\n",
    "    num_cells = len(count_df['unique_id'].unique())\n",
    "    df_ = count_df.groupby(['unique_id','geneID']).size().reset_index(name='counts')\n",
    "    \n",
    "    genes = count_df['geneID'].unique()\n",
    "    cells = count_df['unique_id'].unique()\n",
    "    genes_long = list(genes)*len(cells)\n",
    "\n",
    "    cells_long = []\n",
    "\n",
    "    for c in cells:\n",
    "        cells_long+=[c]*len(genes)\n",
    "    \n",
    "    all_genes_cells = pd.DataFrame({'geneID': genes_long,'unique_id' : cells_long})\n",
    "    df_merge = all_genes_cells.merge(df_,left_on=['geneID','unique_id'],right_on=['geneID','unique_id'],\n",
    "                                how='left').fillna(0)\n",
    "    \n",
    "    count_array = np.zeros((num_cells,num_genes))\n",
    "    \n",
    "    for i,ID in enumerate(count_df['unique_id'].unique()):\n",
    "        counts_ = df_merge[df_merge['unique_id']==ID]\n",
    "        count_array[i,:] = counts_['counts']\n",
    "    \n",
    "        \n",
    "    return(count_array)\n",
    "\n",
    "\n",
    "def get_adjusted_count_matrix(count_df):\n",
    "    ''' Return cell by gene count matrix with intronic counts adjusted by intensity value.\n",
    "    '''\n",
    "    \n",
    "    count_df['adjusted_intron_counts'] = np.ceil(count_df['intensity'].values)\n",
    "\n",
    "    num_genes = len(count_df['geneID'].unique())\n",
    "    num_cells = len(count_df['unique_id'].unique())\n",
    "    df_ = count_df.groupby(['unique_id','geneID'])['adjusted_intron_counts'].sum().reset_index(name='adjusted_intron_counts')\n",
    "    \n",
    "    genes = count_df['geneID'].unique()\n",
    "    cells = count_df['unique_id'].unique()\n",
    "    genes_long = list(genes)*len(cells)\n",
    "\n",
    "    cells_long = []\n",
    "\n",
    "    for c in cells:\n",
    "        cells_long+=[c]*len(genes)\n",
    "    \n",
    "    all_genes_cells = pd.DataFrame({'geneID': genes_long,'unique_id' : cells_long})\n",
    "    df_merge = all_genes_cells.merge(df_,left_on=['geneID','unique_id'],right_on=['geneID','unique_id'],\n",
    "                                how='left').fillna(0)\n",
    "    \n",
    "    count_array = np.zeros((num_cells,num_genes))\n",
    "    \n",
    "    for i,ID in enumerate(count_df['unique_id'].unique()):\n",
    "        counts_ = df_merge[df_merge['unique_id']==ID]\n",
    "        count_array[i,:] = counts_['adjusted_intron_counts']\n",
    "    \n",
    "        \n",
    "    return(count_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in data\n",
    "count_dir = '/home/tara/maria/scBIVI/review/data/takei'\n",
    "es_intron_raw = pd.read_csv(os.path.join(count_dir,'E14-replicate1-intron-seqFISH-ch2.csv'))\n",
    "es_exon_raw = pd.read_csv(os.path.join(count_dir,'E14-replicate1-mRNA-seqFISH-ch1.csv'))\n",
    "\n",
    "es_intron_raw2 = pd.read_csv(os.path.join(count_dir,'E14-replicate2-intron-seqFISH-ch2.csv'))\n",
    "es_exon_raw2 = pd.read_csv(os.path.join(count_dir,'E14-replicate2-mRNA-seqFISH-ch1.csv'))\n",
    "\n",
    "\n",
    "# add unique ids\n",
    "# each field of fov contains different cells, but cell indexing begins again at each fov\n",
    "es_intron_raw['unique_id'] = 'cell' + es_intron_raw['cellID'].astype(str)+'_fov'+es_intron_raw['fov'].astype('str')\n",
    "es_exon_raw['unique_id'] = 'cell' + es_exon_raw ['cellID'].astype(str)+'_fov'+es_exon_raw['fov'].astype('str')\n",
    "es_intron_raw2['unique_id'] = 'cell' + es_intron_raw2['cellID'].astype(str)+'_fov'+es_intron_raw2['fov'].astype('str')\n",
    "es_exon_raw2['unique_id'] = 'cell' + es_exon_raw2['cellID'].astype(str)+'_fov'+es_exon_raw2['fov'].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get burst sizes for all introns\n",
    "es_burst_sizes_all_1 = get_burst_size_per_cell(es_intron_raw)\n",
    "es_burst_sizes_all_2 = get_burst_size_per_cell(es_intron_raw2)\n",
    "\n",
    "# get adjusted intronic counts -- adjust the intron spot by its intensity rounded to an integer\n",
    "es_adjusted_intron_counts1_all = get_adjusted_count_matrix(es_intron_raw)\n",
    "es_adjusted_intron_counts2_all = get_adjusted_count_matrix(es_intron_raw2)\n",
    "\n",
    "# get intronic counts -- should only be ONE or TWO\n",
    "es_intron_counts1_all = get_count_matrix(es_intron_raw)\n",
    "es_intron_counts2_all = get_count_matrix(es_intron_raw2)\n",
    "\n",
    "# get exonic counts\n",
    "es_exon_counts1_all = get_count_matrix(es_exon_raw)\n",
    "es_exon_counts2_all = get_count_matrix(es_exon_raw2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "es_bs_dict_1 = {'burst_size' : es_burst_sizes_all_1,\n",
    "           'gene_name' : es_intron_raw['geneID'].unique(),\n",
    "           'cell_id' : es_intron_raw['unique_id'].unique(),\n",
    "           'adj_intron_counts' : es_adjusted_intron_counts1_all,\n",
    "           'intron_counts' : es_intron_counts1_all, \n",
    "           'exon_counts' : es_exon_counts1_all,\n",
    "           'cell_id_exon' : es_exon_raw['unique_id'].unique(),\n",
    "           'gene_name_exon': es_exon_raw['geneID'].unique() }\n",
    "\n",
    "es_bs_dict_2 = {'burst_size' : es_burst_sizes_all_2,\n",
    "           'gene_name' : es_intron_raw2['geneID'].unique(),\n",
    "           'adj_intron_counts' : es_adjusted_intron_counts2_all,\n",
    "           'intron_counts' : es_intron_counts2_all, \n",
    "           'exon_counts' : es_exon_counts2_all, \n",
    "           'cell_id_exon' : es_exon_raw2['unique_id'].unique(),\n",
    "           'gene_name_exon': es_exon_raw2['geneID'].unique() }\n",
    "\n",
    "# save! \n",
    "with open('./data/takei/E14_rep1_seqFISH_burst_sizes_all','wb') as file:\n",
    "    pickle.dump(es_bs_dict_1,file,protocol=pickle.HIGHEST_PROTOCOL)\n",
    "with open('./data/takei/E14_rep2_seqFISH_burst_sizes_all','wb') as file:\n",
    "    pickle.dump(es_bs_dict_2,file,protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-16-4ebbbfa2ec9d>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-16-4ebbbfa2ec9d>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    g for g in glob.glob('./fits/mc_240521_026_allen_B08_1/*') if '/allen_B08' in g\u001b[0m\n\u001b[0m      ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
